### ISSUE TITLE
display_skipped_hosts flag delays task banner which is confusing on slow tasks

### ISSUE THREAD

**Issue by wasserman** [ansible/ansible#51042]
##### SUMMARY
I have a long running shell script that my playbook runs.  It takes a couple of minutes.  The task immediately before it displayed its banner but it appeared to hang on this task.  After troubleshooting the issue I determined it was the following shell task that gave this apparent hang.  If the task banner of the shell task displayed when it started it would have been obvious which task it was waiting on. 

I discovered this issue with the yaml stdout_plugin.  After trying many other plugins I realized that unixy was still printing the banner at task start.  Since yaml inherits from default I determined this must be why it does this in many others.

Hopefully there is a way to determine if the task will be skipped at task start instead of delaying it.

##### ISSUE TYPE
display_skipped_hosts was set to False in ansible.cfg.  Although this is desired, it was unclear that a consequence of this is the delay of task banners.

##### COMPONENT NAME
lib/ansible/plugins/callback/default.py

##### ANSIBLE VERSION

```ansible 2.7.5
  config file = /home/ec2-user/ansible.cfg
  configured module search path = [u'/home/ec2-user/library', u'/home/ec2-user/apps/library', u'/home/ec2-user/ansible_commands']
  ansible python module location = /home/ec2-user/.local/lib/python2.7/site-packages/ansible
  executable location = /home/ec2-user/.local/bin/ansible
  python version = 2.7.14 (default, Jul 26 2018, 19:59:38) [GCC 7.3.1 20180303 (Red Hat 7.3.1-5)]

```

##### CONFIGURATION

```ACTION_WARNINGS(/home/ec2-user/ansible.cfg) = False
ANSIBLE_FORCE_COLOR(/home/ec2-user/ansible.cfg) = True
COMMAND_WARNINGS(/home/ec2-user/ansible.cfg) = True
DEFAULT_ASK_PASS(/home/ec2-user/ansible.cfg) = False
DEFAULT_BECOME(/home/ec2-user/ansible.cfg) = True
DEFAULT_BECOME_EXE(/home/ec2-user/ansible.cfg) = sudo -u
DEFAULT_BECOME_METHOD(/home/ec2-user/ansible.cfg) = sudo
DEFAULT_BECOME_USER(/home/ec2-user/ansible.cfg) = root
DEFAULT_FORKS(/home/ec2-user/ansible.cfg) = 50
DEFAULT_GATHERING(/home/ec2-user/ansible.cfg) = smart
DEFAULT_HASH_BEHAVIOUR(/home/ec2-user/ansible.cfg) = merge
DEFAULT_MODULE_PATH(/home/ec2-user/ansible.cfg) = [u'/home/ec2-user/library', u'/home/ec2-user/apps/library', u'/home/ec2-user/ansible_commands']
DEFAULT_NO_LOG(/home/ec2-user/ansible.cfg) = False
DEFAULT_STDOUT_CALLBACK(/home/ec2-user/ansible.cfg) = yaml
DEFAULT_TIMEOUT(/home/ec2-user/ansible.cfg) = 30
DEFAULT_UNDEFINED_VAR_BEHAVIOR(/home/ec2-user/ansible.cfg) = True
DISPLAY_SKIPPED_HOSTS(/home/ec2-user/ansible.cfg) = False
HOST_KEY_CHECKING(/home/ec2-user/ansible.cfg) = False
INVENTORY_ENABLED(/home/ec2-user/ansible.cfg) = [u'yaml']
SYSTEM_WARNINGS(/home/ec2-user/ansible.cfg) = True

```

##### OS / ENVIRONMENT

CentOS 7

##### STEPS TO REPRODUCE

```yaml
- name: Illustrating problem
  hosts: all
  tasks:
    - name: Do nothing
      debug: msg="nothing to see here"
    - name: Catch some Z's
      command: sleep 100
```

##### EXPECTED RESULTS

Display the TASK [Catch some Z's] and then hang there.

##### ACTUAL RESULTS

Displays TASK [Do nothing] with output and then hangs without ever showing TASK [Catch some Z's] unless you are patient enough to see if it ever completes.

**Comment by ansibot** [comment]
Files identified in the description:
* [lib/ansible/plugins/callback/default.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/callback/default.py)
* [lib/ansible/plugins/netconf/default.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/netconf/default.py)

If these files are inaccurate, please update the `component name` section of the description or use the `!component` bot command.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
cc @NilashishC @Qalthos @ganeshrn @gdpak @justjais @rcarrillocruz @trishnaguha
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
Files identified in the description:
* [lib/ansible/plugins/action/netconf.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/netconf.py)
* [lib/ansible/plugins/connection/netconf.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/connection/netconf.py)
* [lib/ansible/plugins/netconf/__init__.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/netconf/__init__.py)
* [lib/ansible/plugins/netconf/ce.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/netconf/ce.py)
* [lib/ansible/plugins/netconf/default.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/netconf/default.py)
* [lib/ansible/plugins/netconf/iosxr.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/netconf/iosxr.py)
* [lib/ansible/plugins/netconf/junos.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/netconf/junos.py)
* [lib/ansible/plugins/netconf/sros.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/netconf/sros.py)

If these files are inaccurate, please update the `component name` section of the description or use the `!component` bot command.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
cc @wisotzky
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by wasserman** [comment]
I don't understand what the bot is doing, but I believe the line of code causing this is at https://github.com/ansible/ansible/blob/stable-2.7/lib/ansible/plugins/callback/default.py#L192

**Comment by mkrizek** [comment]
Related to https://github.com/ansible/ansible/issues/49114, possibly a duplicate.

**Comment by wasserman** [comment]
I agree that this is a duplicate.  I looked for this issue and didn't find that one until you referenced it.  Hopefully my example makes it easier to reproduce since it appears there was a challenge with that in the other ticket.
Another idea is that when in DEBUG mode or extra verbose levels, maybe this feature should be ignored since knowing skipped tasks may be important when troubleshooting issues regardless of the ansible.cfg setting.  Just another thought, although this would be a short-term solution since this behavior is still questionable.

**Comment by ansibot** [comment]
Files identified in the description:
* [lib/ansible/plugins/callback/default.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/callback/default.py)

If these files are inaccurate, please update the `component name` section of the description or use the `!component` bot command.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by PikachuEXE** [comment]
#49114 closed as duplicate of this issue

**Comment by PikachuEXE** [comment]
Any update on this issue?
Will this be fixed?

**Comment by bcoca** [comment]
There are conflicting requirements here, as it was a requested feature that we DO NOT display the task header if all hosts are skipped, which we cannot determine until the task 'returns' ... which is how we know if it was skipped or not.

**Comment by PikachuEXE** [comment]
When there are 2 conflicting requirements
Adding an option would be great
Otherwise that change is just scarifying one group of user for another group

**Comment by wasserman** [comment]
@bcoca A skip is quick vs all the non-skips that take a varying amount of time.  This setting is disrupting the ability to debug, especially in situations where one isn't even aware that it exists and that it is set.  I only learned about this feature when trying to use an existing Ansible repo with the included ansible.cfg file.

Couldn't the task_start callback access the task conditional evaluation to improve the logic that calls print_task_banner?  The code has to know IF a task will run before it actually runs it.  Then the first non-skipped host can trigger the banner to print.

**Comment by bcoca** [comment]
@PikachuEXE and that might well be the solution, I was just  pointing out why its not just changing the behaviour.

@wasserman no, since the callback does not have the needed information (nor should it) nor does the invoker of the 'task start' callback have that information either, that gets assembled later on in the process. Also an action itself can decide to 'skip' processing with data not available on the controller, so even if we looked at the task's conditional, that is no guarantee.

**Comment by jamescassell** [comment]
I think checking the conditional should be good enough for most cases, though it might miss some cases and cause extra output.
